app/questionnaire/navigator.py

```diff
@@ -250,10 +256,102 @@ def get_latest_location(self, completed_blocks=None):
         :param completed_blocks:
         :return:
         """
+        location_path = self.get_location_path()
         if completed_blocks:
-            incomplete_blocks = [item for item in self.location_path if item not in completed_blocks]
+            incomplete_blocks = [item for item in location_path if item not in completed_blocks]
 
             if incomplete_blocks:
                 return incomplete_blocks[0]
 
-        return self.first_location
+        return location_path[0]
+
+    def get_front_end_navigation(self, completed_blocks, current_group_id, current_group_instance):
+        """
+        Returns the frontend navigation based on the completed blocks, group id and group instance
+
+        :param completed_blocks:
+        :param group_id:
+        :param group_instance:
+        :return: navigation
+        """
+
+        if 'navigation' not in self.survey_json:
+            return None
+
+        navigation = []
+
+        for group in filter(lambda x: 'hide_in_navigation' not in x, self.survey_json['groups']):
+
+            logger.debug("Building frontend navigation for group %s", group)
+
+            repeating_rule = SchemaHelper.get_repeating_rule(group)
+            completed_id = group['completed_id'] if 'completed_id' in group else group['blocks'][-1]['id']
+
+            if repeating_rule:
+                no_of_repeats = evaluate_repeat(repeating_rule, self.answer_store)
+                link_names = self._generate_link_names(repeating_rule)
+
+                if repeating_rule['type'] == 'answer_count':
+                    self._add_repeating_navigation_item(link_names, completed_blocks, completed_id, group, current_group_id,
+                                                        current_group_instance, navigation, no_of_repeats)
+                elif link_names and no_of_repeats > 0:
```

Comment:
Can `link_names` be removed from the logic here, as it's not used to create the navigation item, and the `no_of_repeats > 0` check should suffice? If this can be removed, we should also move the creation of `link_names` above into the `if repeating_rule['type'] == 'answer_count'` block.